File: app/src/main/java/ie/napkin/supertasks/data/filter/Filter.kt
package ie.napkin.supertasks.data.filter

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Serializable filter tree stored in smart_list_def.filter_json.
 * A small query-compiler ([FilterCompiler]) turns this into SQL, so the UI only edits JSON.
 */
@Serializable
sealed interface Filter {

    @Serializable
    @SerialName("all")
    data class All(val filters: List<Filter>) : Filter

    @Serializable
    @SerialName("any")
    data class AnyOf(val filters: List<Filter>) : Filter

    @Serializable
    @SerialName("not")
    data class Not(val filter: Filter) : Filter

    /** Matches node.done. */
    @Serializable
    @SerialName("done")
    data class Done(val value: Boolean) : Filter

    /**
     * Matches node.in_progress — the task glyph's middle state, the one you swipe to set.
     *
     * It was storable and drawable but not askable: nothing in the query layer knew it existed, so
     * "what am I in the middle of" could be seen and never gathered. A state the app renders but
     * cannot reason about is decoration.
     */
    @Serializable
    @SerialName("in_progress")
    data class InProgress(val value: Boolean) : Filter

    /** Matches node.type; smart lists normally filter for tasks. */
    @Serializable
    @SerialName("type")
    data class Type(val value: String) : Filter

    /**
     * Matches a typed property value. Exactly one of [text]/[number]/[date]/[bool]/[dateRel]
     * should be set (none for IS_SET / NOT_SET). [dateRel] is resolved to epoch millis when
     * the query is compiled, so "due today" style lists stay correct across days.
     */
    @Serializable
    @SerialName("prop")
    data class Prop(
        val defId: String,
        val op: Op,
        val text: String? = null,
        val number: Double? = null,
        val date: Long? = null,
        val bool: Boolean? = null,
        val dateRel: DateRel? = null,
    ) : Filter

    /** Matches nodes tagged with the given label. "Doesn't have" is `Not(HasLabel(id))`. */
    @Serializable
    @SerialName("has_label")
    data class HasLabel(val labelId: String) : Filter

    /**
     * Matches everything that came from one repo — `node.workspace_id`, asked directly.
     *
     * Provenance was filterable before this existed, but only by borrowing [HasLabel]: the reader
     * derived a label per workspace and attached it to every task, so "everything from Work" could
     * be said in the vocabulary that already had a query behind it. That worked and cost more than
     * it looked. The label was a second copy of a column every one of those rows already carried,
     * it shared a namespace with the tags people type — a workspace called `v2-tasks` and a
     * `#v2-tasks` tag are one row, which is unique on (workspace_id, name) — and being a label made
     * it look detachable, so tapping the chip it drew rewrote the file to remove a tag that was
     * never in it. It was also attached to tasks only, while the column is on lists and blocks too.
     *
     * Asking the column is the whole feature with none of that. Note this is a *filter*, distinct
     * from [FilterCompiler.compile]'s `workspaceId`, which fences a rule into the repo that wrote
     * it. This one is a clause a rule may state about itself, and composes: an unscoped Today can
     * say `Any(InWorkspace(a), InWorkspace(b))`, which a fence cannot express.
     */
    @Serializable
    @SerialName("in_workspace")
    data class InWorkspace(val workspaceId: String) : Filter
}

/**
 * Every workspace this rule names, so a view can tell whether it can actually answer itself.
 *
 * A smart list lives in Personal and its clauses may name any repo, which means a device that has
 * not added one of them will match nothing for it — and a Today quietly missing half your tasks is
 * worse than no Today. Nothing here can fix that; the point is to make it sayable.
 *
 * An empty result means the rule names no workspace at all, which is not a rule that spans none —
 * it is a rule that spans every repo this device has, and always answers fully.
 */
fun Filter.workspacesNamed(): Set<String> = buildSet { collectWorkspaces(this) }

private fun Filter.collectWorkspaces(into: MutableSet<String>) {
    when (this) {
        is Filter.InWorkspace -> into += workspaceId
        is Filter.All -> filters.forEach { it.collectWorkspaces(into) }
        is Filter.AnyOf -> filters.forEach { it.collectWorkspaces(into) }
        is Filter.Not -> filter.collectWorkspaces(into)
        // Deliberately exhaustive rather than an else: a new clause that can carry a workspace has
        // to come here too, and the compiler is the only thing that will remember to say so.
        is Filter.Done, is Filter.InProgress, is Filter.Type, is Filter.Prop, is Filter.HasLabel -> Unit
    }
}

@Serializable
enum class Op {
    @SerialName("eq") EQ,
    @SerialName("neq") NEQ,
    @SerialName("lt") LT,
    @SerialName("lte") LTE,
    @SerialName("gt") GT,
    @SerialName("gte") GTE,
    @SerialName("is_set") IS_SET,
    @SerialName("not_set") NOT_SET,
}

@Serializable
enum class DateRel {
    @SerialName("today_start") TODAY_START,
    @SerialName("today_end") TODAY_END,
}

@Serializable
data class SortSpec(
    val by: SortBy,
    val defId: String? = null,   // required for the PROP_* kinds
    val desc: Boolean = false,
    val nullsLast: Boolean = true,
)

@Serializable
enum class SortBy {
    @SerialName("prop_date") PROP_DATE,
    @SerialName("prop_number") PROP_NUMBER,
    @SerialName("prop_text") PROP_TEXT,
    @SerialName("title") TITLE,
    @SerialName("created") CREATED,
}

/**
 * One property value to auto-set on tasks created inside a smart list (write side).
 * [dateRel] defers date resolution to insert time ("due today" lists stamp the actual
 * today) — see SmartListRepository.addTask.
 */
@Serializable
data class ApplyOnCreate(
    val defId: String,
    val text: String? = null,
    val number: Double? = null,
    val date: Long? = null,
    val bool: Boolean? = null,
    val dateRel: DateRel? = null,
)

val FilterJson: Json = Json {
    ignoreUnknownKeys = true
    encodeDefaults = false
    classDiscriminator = "kind"
}

/**
 * Per the spec: equality clauses are writable and self-satisfying, so the write-side
 * apply_on_create values are derived straight from the filter's EQ clauses. Two extensions:
 *  - today-relative comparisons ("due today or earlier/later") are satisfied by Due = today,
 *    emitted as a deferred [ApplyOnCreate.dateRel] resolved at insert time;
 *  - an AnyOf is satisfied by satisfying one branch — the first derivable branch wins.
 */
fun deriveApplyOnCreate(filter: Filter): List<ApplyOnCreate> = when (filter) {
    is Filter.All -> filter.filters.flatMap { deriveApplyOnCreate(it) }
    is Filter.AnyOf -> filter.filters.firstNotNullOfOrNull { branch ->
        deriveApplyOnCreate(branch).takeIf { it.isNotEmpty() }
    } ?: emptyList()
    is Filter.Prop -> when {
        filter.op == Op.EQ && filter.dateRel == null -> listOf(
            ApplyOnCreate(
                defId = filter.defId,
                text = filter.text,
                number = filter.number,
                date = filter.date,
                bool = filter.bool,
            )
        )
        (filter.op == Op.LTE && filter.dateRel == DateRel.TODAY_END) ||
            (filter.op == Op.GTE && filter.dateRel == DateRel.TODAY_START) -> listOf(
            ApplyOnCreate(defId = filter.defId, bool = false, dateRel = DateRel.TODAY_START)
        )
        else -> emptyList()
    }
    else -> emptyList() // not/other comparisons can't be auto-matched -> read-mostly
}

/**
 * Rewrites an "open tasks" filter into its completed counterpart by flipping every
 * [Filter.Done] clause, so the same rules can be asked the opposite question: not "what is
 * still due today" but "what due today did I finish". Returns null when the tree has no Done
 * clause to flip — such a list already admits completed tasks, so there is no separate
 * completed set to fetch.
 */
fun completedVariant(filter: Filter): Filter? =
    // A list of started tasks has no completed half: finishing a task clears in_progress in the
    // same UPDATE, so flipping the done clause would ask for a set that cannot exist and render an
    // empty "DONE" heading under every such view.
    if (!hasDoneClause(filter) || asksForStarted(filter)) null else flipDone(filter)

private fun asksForStarted(f: Filter): Boolean = when (f) {
    is Filter.InProgress -> f.value
    is Filter.All -> f.filters.any { asksForStarted(it) }
    is Filter.AnyOf -> f.filters.any { asksForStarted(it) }
    is Filter.Not -> false
    else -> false
}

private fun hasDoneClause(f: Filter): Boolean = when (f) {
    is Filter.Done -> !f.value
    is Filter.All -> f.filters.any { hasDoneClause(it) }
    is Filter.AnyOf -> f.filters.any { hasDoneClause(it) }
    is Filter.Not -> hasDoneClause(f.filter)
    else -> false
}

private fun flipDone(f: Filter): Filter = when (f) {
    is Filter.Done -> Filter.Done(!f.value)
    is Filter.All -> Filter.All(f.filters.map { flipDone(it) })
    is Filter.AnyOf -> Filter.AnyOf(f.filters.map { flipDone(it) })
    is Filter.Not -> Filter.Not(flipDone(f.filter))
    else -> f
}

File: ios/YantraCore/Sources/YantraCore/Filter.swift
import Foundation

/// The smart-list filter tree — `data/filter/Filter.kt`. JSON uses `kind` as the discriminator and
/// omits defaults, exactly as kotlinx.serialization writes it, so a rule made on either phone is
/// the other's to read.
public indirect enum Filter: Equatable, Sendable {
    case all([Filter])
    case anyOf([Filter])
    case not(Filter)
    case done(Bool)
    case inProgress(Bool)
    case type(String)
    case prop(defId: String, op: Op, text: String? = nil, number: Double? = nil, date: Int64? = nil, bool: Bool? = nil, dateRel: DateRel? = nil)
    case hasLabel(String)
    case inWorkspace(String)

    public var workspacesNamed: Set<String> {
        switch self {
        case let .inWorkspace(w): return [w]
        case let .all(fs), let .anyOf(fs): return fs.reduce(into: []) { $0.formUnion($1.workspacesNamed) }
        case let .not(f): return f.workspacesNamed
        default: return []
        }
    }
}

public enum Op: String, Codable, Sendable { case eq, neq, lt, lte, gt, gte, isSet = "is_set", notSet = "not_set" }
public enum DateRel: String, Codable, Sendable { case todayStart = "today_start", todayEnd = "today_end" }
public enum SortBy: String, Codable, Sendable { case propDate = "prop_date", propNumber = "prop_number", propText = "prop_text", title, created }

public struct SortSpec: Codable, Equatable, Sendable {
    public var by: SortBy
    public var defId: String?
    public var desc: Bool
    public var nullsLast: Bool
    public init(by: SortBy, defId: String? = nil, desc: Bool = false, nullsLast: Bool = true) {
        self.by = by; self.defId = defId; self.desc = desc; self.nullsLast = nullsLast
    }
    enum CodingKeys: String, CodingKey { case by, defId, desc, nullsLast }
    public init(from d: Decoder) throws {
        let c = try d.container(keyedBy: CodingKeys.self)
        by = try c.decode(SortBy.self, forKey: .by)
        defId = try c.decodeIfPresent(String.self, forKey: .defId)
        desc = try c.decodeIfPresent(Bool.self, forKey: .desc) ?? false
        nullsLast = try c.decodeIfPresent(Bool.self, forKey: .nullsLast) ?? true
    }
    public func encode(to e: Encoder) throws {
        var c = e.container(keyedBy: CodingKeys.self)
        try c.encode(by, forKey: .by)
        try c.encodeIfPresent(defId, forKey: .defId)
        if desc { try c.encode(desc, forKey: .desc) }
        if !nullsLast { try c.encode(nullsLast, forKey: .nullsLast) }
    }
}

public struct ApplyOnCreate: Codable, Equatable, Sendable {
    public var defId: String
    public var text: String?
    public var number: Double?
    public var date: Int64?
    public var bool: Bool?
    public var dateRel: DateRel?
    public init(defId: String, text: String? = nil, number: Double? = nil, date: Int64? = nil, bool: Bool? = nil, dateRel: DateRel? = nil) {
        self.defId = defId; self.text = text; self.number = number; self.date = date; self.bool = bool; self.dateRel = dateRel
    }
}

public enum BuiltIns {
    public static let priority = "builtin-priority", due = "builtin-due", deadline = "builtin-deadline", assignee = "builtin-assignee"
}

// MARK: Codable with the `kind` discriminator

extension Filter: Codable {
    enum K: String, CodingKey { case kind, filters, filter, value, defId, op, text, number, date, bool, dateRel, labelId, workspaceId }

    public init(from d: Decoder) throws {
        let c = try d.container(keyedBy: K.self)
        switch try c.decode(String.self, forKey: .kind) {
        case "all": self = .all(try c.decode([Filter].self, forKey: .filters))
        case "any": self = .anyOf(try c.decode([Filter].self, forKey: .filters))
        case "not": self = .not(try c.decode(Filter.self, forKey: .filter))
        case "done": self = .done(try c.decode(Bool.self, forKey: .value))
        case "in_progress": self = .inProgress(try c.decode(Bool.self, forKey: .value))
        case "type": self = .type(try c.decode(String.self, forKey: .value))
        case "prop": self = .prop(defId: try c.decode(String.self, forKey: .defId), op: try c.decode(Op.self, forKey: .op),
                                 text: try c.decodeIfPresent(String.self, forKey: .text), number: try c.decodeIfPresent(Double.self, forKey: .number),
                                 date: try c.decodeIfPresent(Int64.self, forKey: .date), bool: try c.decodeIfPresent(Bool.self, forKey: .bool),
                                 dateRel: try c.decodeIfPresent(DateRel.self, forKey: .dateRel))
        case "has_label": self = .hasLabel(try c.decode(String.self, forKey: .labelId))
        case "in_workspace": self = .inWorkspace(try c.decode(String.self, forKey: .workspaceId))
        case let k: throw DecodingError.dataCorruptedError(forKey: .kind, in: c, debugDescription: "unknown filter kind \(k)")
        }
    }

    public func encode(to e: Encoder) throws {
        var c = e.container(keyedBy: K.self)
        switch self {
        case let .all(fs): try c.encode("all", forKey: .kind); try c.encode(fs, forKey: .filters)
        case let .anyOf(fs): try c.encode("any", forKey: .kind); try c.encode(fs, forKey: .filters)
        case let .not(f): try c.encode("not", forKey: .kind); try c.encode(f, forKey: .filter)
        case let .done(v): try c.encode("done", forKey: .kind); try c.encode(v, forKey: .value)
        case let .inProgress(v): try c.encode("in_progress", forKey: .kind); try c.encode(v, forKey: .value)
        case let .type(v): try c.encode("type", forKey: .kind); try c.encode(v, forKey: .value)
        case let .prop(defId, op, text, number, date, bool, dateRel):
            try c.encode("prop", forKey: .kind); try c.encode(defId, forKey: .defId); try c.encode(op, forKey: .op)
            try c.encodeIfPresent(text, forKey: .text); try c.encodeIfPresent(number, forKey: .number); try c.encodeIfPresent(date, forKey: .date)
            try c.encodeIfPresent(bool, forKey: .bool); try c.encodeIfPresent(dateRel, forKey: .dateRel)
        case let .hasLabel(id): try c.encode("has_label", forKey: .kind); try c.encode(id, forKey: .labelId)
        case let .inWorkspace(w): try c.encode("in_workspace", forKey: .kind); try c.encode(w, forKey: .workspaceId)
        }
    }
}

/// kotlinx-compatible JSON: compact, keys in declaration order (JSONEncoder does not promise order,
/// so smart-list definitions written by iOS are re-read by Android fine but may not be byte-equal;
/// the plan is to route them through `JSONValue` before the first shared write).
public enum FilterJSON {
    public static func encode<T: Encodable>(_ v: T) -> String {
        let e = JSONEncoder(); e.outputFormatting = [.withoutEscapingSlashes]
        return String(decoding: try! e.encode(v), as: UTF8.self)
    }
    public static func decode<T: Decodable>(_ t: T.Type, _ s: String) -> T? {
        try? JSONDecoder().decode(t, from: Data(s.utf8))
    }
}

// MARK: derive apply-on-create (write side)

/// Equality clauses are writable and self-satisfying — `FilterCompiler.deriveApplyOnCreate`.
public func deriveApplyOnCreate(_ f: Filter) -> [ApplyOnCreate] {
    switch f {
    case let .all(fs): return fs.flatMap(deriveApplyOnCreate)
    case let .anyOf(fs): return fs.lazy.map(deriveApplyOnCreate).first { !$0.isEmpty } ?? []
    case let .prop(defId, op, text, number, date, bool, dateRel):
        if op == .eq, dateRel == nil { return [ApplyOnCreate(defId: defId, text: text, number: number, date: date, bool: bool)] }
        if (op == .lte && dateRel == .todayEnd) || (op == .gte && dateRel == .todayStart) {
            return [ApplyOnCreate(defId: defId, bool: false, dateRel: .todayStart)]
        }
        return []
    default: return []
    }
}

/// The flipped rule for the "done" half of a view, or nil when the rule has no `done(false)` or asks
/// for in-progress tasks (finishing clears that flag, so the half cannot exist).
public func completedVariant(_ f: Filter) -> Filter? {
    func has(_ f: Filter, _ p: (Filter) -> Bool) -> Bool {
        if p(f) { return true }
        switch f {
        case let .all(fs), let .anyOf(fs): return fs.contains { has($0, p) }
        case let .not(x): return has(x, p)
        default: return false
        }
    }
    guard has(f, { if case .done(false) = $0 { return true }; return false }),
          !has(f, { if case .inProgress(true) = $0 { return true }; return false }) else { return nil }
    func flip(_ f: Filter) -> Filter {
        switch f {
        case let .done(v): return .done(!v)
        case let .all(fs): return .all(fs.map(flip))
        case let .anyOf(fs): return .anyOf(fs.map(flip))
        case let .not(x): return .not(flip(x))
        default: return f
        }
    }
    return flip(f)
}
